llama.cpp

```diff
@@ -12537,7 +12582,8 @@ struct llama_context_params llama_context_default_params() {
     struct llama_context_params result = {
         /*.seed                        =*/ LLAMA_DEFAULT_SEED,
         /*.n_ctx                       =*/ 512,
-        /*.n_batch                     =*/ 512,
+        /*.n_batch                     =*/ 4096,
```

Comment:
A bigger batch size is what allows pipeline parallelism to work. For example if the application submits a batch of 4096 tokens to `llama_decode`, these will be split in mini-batches of 256 tokens each (`n_ubatch`) and evaluated in as a pipeline in parallel between the available GPUs. It could be reduced to 2048 and still get most of the benefit.

Reply:
Pipeline parallelism by default seems desirable. But `n_batch` shouldn't exceed `n_ctx`. Even when passing `--ctx-size 128` to `main`, `n_batch` is still 4096 (from the default), and the `785.63 MiB` of logits are still allocated, even if they can't be used since a batch bigger than `n_ctx` will simply not be able to find a big enough KV slot (for Transformer-based models, at least).

Reply:
We can change `llama_new_context_with_model` to limit `n_batch` and `n_ubatch` to `n_ctx`, since there is no advantage to increasing it beyond that.

Ideally, we would set the defaults intelligently according to the hardware of the system, and only increase `n_batch` is the system can actually support pipeline parallelism, which requires several CUDA GPUs and full offload. However that would require deeper changes.

Reply:
> But `n_batch` shouldn't exceed `n_ctx`

This is not always the case. When using a model that does not utilize the KV cache (for example non-causal embedding model like BERT), we might want to run with `n_ctx = 0, n_batch = 8192`

With [4400153](https://github.com/ggerganov/llama.cpp/pull/6017/commits/4400153348693db3ef008dbf73c78a2a3ae6e39d) applied in such cases, we now have to allocate a KV cache due to `n_ctx = 8192` and it won't be used

Given that, should we revert the clamp change?

Reply:
I didn't realize that, I also added a check in `llama_new_context_with_model` that prevents `n_ctx` being zero. It would be good to clamp the value if possible to avoid wasting memory. Is there a simple way to check for this? Maybe `hparams.casual_attn`?

Reply:
`n_ctx = 0` actually would default to `n_ctx_train` which is OK.
By `n_ctx = 0` I meant to be some small value (e.g. `n_ctx = 16`) in order to not waste VRAM on the KV cache when it is not used

Yes, we can check `hparams.causal_attn`. When it is `false` then the graph should not use the KV cache.

Btw, note that we now have support for a special model GritLM that has `hparams.causal_attn == true`, but it supports at runtime to set `cparams.causal_attn == false` in order to generate embeddings. Currently, we will still reuse the same compute graph both for text generation and for embeddings, so the KV cache is necessary in both situations. Though, in some potential implementation we could build 2 different graphs - one with and one without KV cache.

But to keep things more simple, let's not consider that for now. Checking `hparams.causal_attn` should be good for now and when it is `false` it should imply that the KV cache will not be used by the compute graph of the model

Reply:
> Ideally, we would set the defaults intelligently according to the hardware of the system, and only increase n_batch if the system can actually support pipeline parallelism, which requires several CUDA GPUs and full offload.

Should the default `n_batch` be more conservative then, for now?
For models with large context (e.g when using `-c 16384`), where clamping doesn't reduce `n_batch`, a default `n_batch` of 4096 still uses around `500 MiB` more memory than the previous default of 512. For recurrent models, this is much bigger than the KV cache size.
At least for my use case this means I'll now have to specify `-b 512` to save memory, whereas users with many GPUs probably already set a bigger `n_batch` anyway (I might be wrong, though).

Ideally though, there should be a way to know how many logits per decode will actually be used at most, while still allowing bigger logical batches than that. This would solve the memory usage problem and thus a big default `n_batch` should be fine.
I think only perplexity calculations (in `perplexity` and `imatrix`) require the logits for the whole batch, while normally only as many logits as distinct sequences are required in the other examples. But `llama_get_logits_ith(ctx, i)` would no longer be equivalent to `llama_get_logits(ctx) + i*n_vocab`. I might explore this idea in a future PR.

Reply:
We could pre-allocate buffers for `n_max_seq` tokens only during initialization, and increase the size dynamically in `llama_decode` automatically if there is a request for more logits than that.

Reply:
> We could pre-allocate buffers for `n_max_seq` tokens only during initialization, and increase the size dynamically in `llama_decode` automatically if there is a request for more logits than that.

From what I understand, the logits aren't necessarily contiguous with each other in the output buffer, so yes pre-allocation and dynamic resizing could be done, but not until the way the layout of the logits is always made contiguous with no offset before the first used logits.

Reply:
We could definitely improve the handling of logits. Even `perplexity` and `imatrix` only need logits for `n_ctx/2`. We could also skip the computation of the output layer for the tokens where logits are not needed. IIRC there was a PR about this a while ago, but it was never merged.